Rewrites unarmed attacks on fellow humans

emmanuelbassil opened the pull request:
🆑 emmanuelbassil
tweak: Rewrote unarmed attack code to use the same miss/parry code as melee attacks with weapons.
bugfix: Holo-boxing gloves no longer target the wrong limb 100% of the time.
bugfix: Punching simple animals should make a punching noise again.
bugfix: Fixed cigarette wrapping paper icons, now becomes empty when out of wrapping paper.
bugfix: Rapid pipe deployer no longer creates a pipe as you put it into storage.
/🆑 

Let the record state, this initially started as a PR to just fix the trident wrapping paper icons. I don't know how it turned into this.
Extensively tested, numbers haven't been changed much except get_parry_chance which now gets the difference between skill instead of just the skill of the parrier. Past behavior had parry_chance increase by 10, 20, or 30 if trained/experienced/master. Now it checks difference between attacker/defender and increases parry chance by 5 for each difference in level; to a maximum of 20. 

The old way it was designed, punches used to get a buff if they were 'accurate'. If so; damage always was 5. And given having a damage of 5 triggers things like a punch knocking you over, dizzying you, dropping your weapon, etc; ~~I did not want to have all punches rand(1,5). I kept that functionality in a more elegant way: If resolve_attack returned a different area than what you're targeting, it's considered an inaccurate attack and damage is from 1 to 5. If you land your punch where you intended, it becomes 4,5 meaning 'glancing' punches are not possible if you're accurate. Also increases the chance of special punches triggering. This is still technically a nerf, as in the past code all accurate attacks returned 5. Although nerf is counter-balanced by better accuracy code in general.~~
**The striked part is now inaccurate.** Accurate punches/chance of special attacks will have the exact same chance post-refactor to minimize the number of changes. See comments below.

Also fixed a few bugs in a separate commit.

MuckerMayhem commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -156,13 +156,33 @@ meteor_act
+	var/obj/item/organ/external/affecting = get_organ(hit_zone)
+	if (!affecting || affecting.is_stump())
+		to_chat(user, SPAN_DANGER("They are missing that limb!"))
+		return null
```
There's no need to specifically return null, that's the default returned value if nothing is specified.
```suggestion
		return
	if (check_shields(damage, null, user, target_zone, user.name))
		return

	var/obj/item/organ/external/affecting = get_organ(hit_zone)
	if (!affecting || affecting.is_stump())
		to_chat(user, SPAN_DANGER("They are missing that limb!"))
		return
```

emmanuelbassil commented:
Fixed.

SierraKomodo reviewed (commented):
I missed hitting the test merge label on this one.

Considering the already in progress testing of #33975, going to request this one waits until after that one's live and fully bug-free.

emmanuelbassil commented:
So I've been asked to try to keep the %ages regarding attack success and strong attack as similar as I can to existing implementation. Below is my dissertation on the matter.

Current implementation, as per the comments left by the original author of punch code, is:
`
					If aiming for chest:
						27.3% chance you hit your target organ
						70.5% chance you hit a random other organ
						 2.2% chance you miss
					If aiming for something else:
						23.2% chance you hit your target organ
						56.8% chance you hit a random other organ
						15.0% chance you miss`

Damage is rand(1,5) + any damage the fist has. Usually 0 for human punches, claws have a bit extra. If an attack is deemed accurate, damage is always turned to 5. In existing implementation, an attack is accurate if:
- Victim is on help intent
- Victim is grabbed or down

Then as the attack is processed, it calls the attack's apply_effects(). This is the proc that causes staggering, being weakened, falling back, blurry vision, your leg buckling, or your arm dropping whatever it's holding depending on which zone is hit. This only triggers if your attack damage is 5 or more and zone armor is 0; there's a 25% chance of processing the special effect and if that fails another 25% chance of causing weakening.

So for non-accurate attacks, a regular human with no hulk/claws has a 20% chance of a full strength punch, which then leads to a 43% chance of a special attack described above. So for any given punch you have a total 8.6% chance of a special punch.
In accurate attacks; conditions described above, it's a 43% chance of a special attack.

///////////////////////

Post-refactor, it will be using a better miss code that is weighted towards how difficult a target is to hit. This will also take shields, parrying, combat skill differential, and whether the attacker is shocked/confused/weakened/blind. In other words, it'll make use of the wealth of miss code that is used for melee attacks with items and apply it to punches. This means that it's difficult for me to calculate every single scenario, which I could do for the above. I could, however, calculate the basics and re-adjust.

Base miss chance in the code is as follows:
`BP_HEAD = 70,
	BP_EYES = 70,
	BP_MOUTH = 70,
	BP_CHEST = 10,
	BP_GROIN = 20,
	BP_L_LEG = 60,
	BP_R_LEG = 60,
	BP_L_ARM = 30,
	BP_R_ARM = 30,
	BP_L_HAND = 50,
	BP_R_HAND = 50,
	BP_L_FOOT = 70,
	BP_R_FOOT = 70`

This number is then taken and combat skill differentials and any other modifiers like shocked/weakened/confused/blind is added to it and run through this:
`if(prob(miss_chance))
		return null
else
	if (prob(miss_chance))
		return (ran_zone())
	else
		return zone`

Of note, if a target is on help intent, buckled, lying down, or grabbed then none of this runs and the zone is returned as is.

So under the refactor, if someone hits the chest with a 10% missing base chance assuming no other modifiers you will have a:
10% chance of missing (Was 2%)
9% chance of hitting another organ ( Was 70%)
81% of hitting the chest itself. (Was 27.3%)

While this seems like a drastic change, this brings punching into line with existing behavior for melee attacks with weapons. Chest is the easiest target to hit, and having a 73% chance of your chest attack missing was wild.

If we do the other extreme, like the head or the foot with a 70% base chance we get:
70% chance of missing outright (Was  15%)
21% chance of hitting another organ (56.8%)
9% chance of striking true (Was 23%)

Again, drastic change but this is already existing behavior for everything but punches. Current implementation of punches makes zone picking useless, where everything but the chest leads to largely the same. I'm not going to tweak any of this, as part of the refactor is bringing miss chances in line with existing code elsewhere. If you decide to target the head or the foot, you must live with the risk involved. On the flip side; chest attacks are MUCH more accurate.

The problem I need to solve is make the %age of a special attack (staggering, dizziness, etc) largely similar to pre-existing implementation; because if that changes drastically there'll be a problem.
Will go over this in a further comment to not make this insanely big.

emmanuelbassil commented:
For special attacks, I am leaving the numbers exactly similar to prior implementation:
Will duplicate code found in get_zone_with_miss_chance() to check if a target is on help, down, grabbed aggressively, or buckled. If so, punch damage will always be 5 which means that any punch on such a target has a 43% chance of triggering a special attack. Likewise, any punch on a target who doesn't meet the above conditions will have an 8.6% chance of special attack per landed punch, even if the punch misses the intended organ and lands elsewhere.

I was going to proceed with an implementation that rewards accurate punches (ergo, hitting the left arm while aiming for the left arm), but that would require lots of number changes which I want to minimize in this PR. Effectively, the **only change in %ages** is the change in miss chances/accuracy, which is only happening because I'm bringing punch code in line with item melee code.

emmanuelbassil committed:
Refactors human-on-human punching

Auto stash before rebase of "refs/heads/antlion"
Auto stash before rebase of "refs/heads/AssortedFixes"
Fixes

Auto stash before rebase of "refs/heads/AssortedFixes"

emmanuelbassil committed:
Assorted bugfixes

Fixes

Fixes

Fixes

emmanuelbassil commented:
Merge conflicts fixed

SierraKomodo reviewed (approved):
There haven't been any issues during the test merge, this one looks good to go.
